doc/src/guides/contributing/deprecations.md

```diff
@@ -0,0 +1,388 @@
+(deprecation-policy)=
+# Deprecation Policy
+
+This page outlines SymPy's police on doing deprecations, and describes the
+steps developers should take to properly deprecate code.
+
+## What is a deprecation?
+
+A deprecation is a way to make backwards incompatible changes in a way that
+allows users to update their code. Deprecated code continues to work as it
+used to, but whenever someone uses it, it prints a warning to the screen
+indicating that it will be removed in a future version of SymPy, and
+indicating what the user should be using instead.
+
+This allows users a chance to update their code without it completely
+breaking. It also gives SymPy an opportunity to give users an informative
+message on how to update their code, rather than making their code simply
+error or start giving wrong answers.
+
+## When to deprecate and when not to
+
+Sometimes the API of SymPy must change in an incompatible way. We try to avoid
+this, because changing API breaks people's code, but it is often deemed worthy
+to do so. Some reasons APIs are changed can include:
+
+- The existing API is confusing.
+- There is unnecessary redundancy in the API.
+- The existing API limits what is possible.
```

Comment:
Looking through the list of currently open issues tagged with deprecation removal it seems that the most common reason is just because a change is being made to something that seems private but it isn't clear whether anyone downstream might be using. The problem is that the distinction between public and private is not clearly delineated anywhere so we end up deprecating things that probably no one cares about.

Reply:
Another reason would be because something is fundamentally broken and cannot be made to work sensibly without incompatible changes.